File: code/ts/src/http/http_client.ts
import { ServiceClientConfig, ServiceRequest, type ServiceClient } from '../core/service_client';
import { HttpClientException } from './http_client_exception';
import { HttpServiceClient } from './http_service_client';

/// Convenience function for one-shot HTTP requests.
///
/// Creates a temporary HttpServiceClient, sends the request,
/// and closes the client. Returns the parsed response data.
///
/// @deprecated Use HttpServiceClient directly. Will be removed in v0.3.0.
export async function httpClient(opts: {
  method: string;
  baseUrl: string;
  endpoint: string;
  headers?: Record<string, string>;
  body?: Record<string, unknown>;
  errorMessage?: string;
}): Promise<unknown> {
  const errorMessage = opts.errorMessage ?? 'Error in HTTP request';
  let client: ServiceClient | undefined;

  const config = new ServiceClientConfig({
    baseUrl: opts.baseUrl,
    defaultHeaders: opts.headers ?? {},
  });

  try {
    client = new HttpServiceClient(config);

    const request = ServiceRequest.http({
      method: opts.method,
      endpoint: opts.endpoint,
      body: opts.body,
      errorMessage,
    });

    const response = await client.send(request);
    return response.data;
  } catch (e) {
    if (e instanceof HttpClientException) throw e;
    throw new Error(`${errorMessage}: [Connection error] - ${e}`);
  } finally {
    await client?.close();
  }
}

File: code/dart/lib/src/http/http_client.dart
import 'dart:developer' as dev;

import 'package:http/http.dart' as http;

import '../core/service_core.dart';
import 'http_exceptions.dart';
import 'http_service_client.dart';

/// Convenience function for one-shot HTTP requests.
///
/// Creates a temporary [HttpServiceClient], sends the request,
/// and closes the client. Returns the parsed response data.
///
/// The optional [client] parameter exists for testing only.
@Deprecated('Use HttpServiceClient directly. Will be removed in v0.3.0.')
Future<dynamic> httpClient({
  required String method,
  required String baseUrl,
  required String endpoint,
  Map<String, String>? headers,
  Map<String, dynamic>? body,
  String errorMessage = 'Error in HTTP request',
  http.Client? httpClient,
}) async {
  ServiceClient? serviceClient;
  final config = ServiceClientConfig(
    baseUrl: Uri.parse(baseUrl),
    defaultHeaders: headers ?? const {},
  );

  try {
    serviceClient = HttpServiceClient(config, client: httpClient);

    final request = ServiceRequest.http(
      method: method,
      endpoint: endpoint,
      body: body,
      errorMessage: errorMessage,
    );

    dev.log('\n🔵 HTTP Request:');
    dev.log('   Method: $method');
    dev.log('   URL: $baseUrl$endpoint');
    if (body != null) {
      dev.log('   Body: $body');
    }

    final response = await serviceClient.send(request);

    dev.log('🟢 HTTP Response:');
    dev.log('   Status: ${response.statusCode}');
    dev.log('   Data: ${response.data}');

    return response.data;
  } on HttpClientException {
    rethrow;
  } catch (e) {
    dev.log('🔴 HTTP Client Error: $e');
    dev.log('   URL: $method $baseUrl$endpoint');
    throw Exception('$errorMessage: [Connection error] - $e');
  } finally {
    await serviceClient?.close();
  }
}
